hw5_iterator/main.py

```diff
@@ -0,0 +1,243 @@
+            else:
+                prev_id = s     # store id-line for the next sequence reading
+                break
+        if not seq:
+            break
+        yield idx, seq
+
+
+def task1_test():
+
+    reader = fasta_reader('sequences.fasta')
+    print(type(reader))
+
+    for id_, seq in reader:
+        print(id_, seq[:50])
+
+
+# -----------------------------------------------------------------------------
+# Task 2 - Class for reading sequences with small changes.
+# -----------------------------------------------------------------------------
+
+class FastaChangingReader:
+    '''
+    This class is iterator that every call reads the next sequence
+    from file, makes small random changes of sequence and returns
+    two values: id and seq.
+    '''
+    def __init__(self, file_path: str):
+        '''
+        Parameter:
+            file_path - path to FASTE file.
+        '''
+        self.file_path = file_path
+        self.gen = fasta_reader(file_path)
+
+    def __iter__(self):
+        return self
+
+    def _make_deletion(self, seq: str) -> str:
+        '''
+        Delete part of sequence in random place with random length.
+        '''
+        start = random.randrange(len(seq))
+        length = random.randint(1, 10)
+        # print(f'start={start}, len={length}')
+        return seq[:start] + seq[start + length:]
+
+    def _make_insertion(self, seq: str) -> str:
+        '''
+        Insert random sub-sequence in random place.
+        To don't make mistake with the letter - get them from current
+        string.
+        '''
+        iseq = ''
+        for i in range(random.randint(1, 10)):
+            iseq += seq[random.randrange(len(seq))]
+        # print(iseq)
+        place = random.randrange(len(seq))
+        return seq[:place] + iseq + seq[place:]
+
+    def _make_replacement(self, seq: str) -> str:
+        '''
+        Replace some sun-sequence with random length in random place
+        with another one sub-sequences with the same length in anothe
+        random place.
+        '''
+        place1 = random.randrange(len(seq))
+        place2 = random.randrange(len(seq))
+        length = random.randint(1, 10)
+        # print(f'p1={place1}, p2={place2}, l={length}')
+        iseq1 = seq[place1: place1 + length]
+        iseq2 = seq[place2: place2 + length]
+        seq = seq[: place1] + iseq2 + seq[place1 + length:]
+        seq = seq[: place2] + iseq1 + seq[place2 + length:]
+        return seq
+
+    def _random_change(self, seq: str) -> str:
+        '''
+        Do small random changing of sequence.
+        '''
+        # print('org:', seq)
+        operation = random.randint(1, 4)
+        if operation == 1:
+            seq = self._make_deletion(seq)
+        elif operation == 2:
+            seq = self._make_insertion(seq)
+        elif operation == 3:
+            seq = self._make_replacement(seq)
+        else:
+            pass  # no changes
```

Comment:
Альтернативный вариант
```suggestion
operation = random.choice([
            self._make_deletion,
            self._make_insertion,
            self._make_replacement
])
seq = operation(seq)
```